### PR TITLE
[Clustering] Independent Clustering Verification

### PR DESCRIPTION
**Pull request by AlexandreSinger**
Created a method that would independently verify the clustering in the VPR flow. If a clustering passes this verification, it is assumed that it can be used in placement and routing without issue.

By design, this method does not use any global variables (everything needs to be passed in) and it recomputes everything that it does not assume. This allows it to be independent of the packing flow, so this method can also be used in the AP flow.

### CODE REVIEW

**Comment by AlexandreSinger** [comment]
@vaughnbetz This is a complementary PR to the Independent Placement Verification PR I raised yesterday (PR #2769 ). I could not find anywhere in the clustering flow that checked these invariants directly so this should improve test coverage. This checks the clustering to ensure that every atom is in a cluster, every cluster has correct floorplanning constraints, and the cluster pbs are correct (for the most part). Please review.

File: vpr/src/pack/verify_clustering.cpp [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+                                                  atom_nlist,
+                                                  atom_lookup);
+    // Check that all clusters are clustering atoms which are constrained to
+    // overlapping regions and the constraints of the clusters are consistent
+    // with the atoms they contain.
+    num_errors += check_clustering_floorplanning_consistency(clb_nlist,
+                                                             clb_atoms,
+                                                             cluster_constraints,
+                                                             constraints);
+    // TODO: There exists more checks for the clustering in base/check_netlist.cpp
```
Comment (vaughnbetz) [inline vpr/src/pack/verify_clustering.cpp:428]: I think general netlist features like mapping consistency are good to put in the netlist verify routine if possible. But I guess given that the atom_lookup clb mapping is not in the netlist or tied to it, it's OK to have the checks here instead.  Just suggesting you think about whether we should have a lookup verifier that can be called by this routine for checks that make sense ...

File: vpr/src/pack/verify_clustering.cpp [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+    }
+    return num_errors;
+}
+
+/**
+ * @brief Check that the floorplanning constraints on the cluster is consistent
+ *        with the floorplanning constraints on the atoms.
+ *
+ * This checks the following invariants:
+ *  - If a cluster is unconstrained, all of its atoms are unconrstrained.
```
Comment (vaughnbetz) [inline vpr/src/pack/verify_clustering.cpp:225]: unconstrained typo

File: vpr/src/pack/verify_clustering.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * @brief Verifies the clustering of the atom netlist as it appears in the
+ *        given context.
+ *
+ * This performs the verification in the method above, but performs it on the
+ * given VPR context itself. This verifies that the actual clustering being used
+ * through the rest of the flow is valid.
+ *
+ * NOTE: The above method is used when one wishes to verify a temporary
```
Comment (vaughnbetz) [inline vpr/src/pack/verify_clustering.h:75]: I'd say the verify_clustering () function with more parameters is used ...
(above can get out of date with code restructuring)

**Review by vaughnbetz — APPROVED** [review]
Looks good; merge when you see fit.

**Pull request merged**